### PR TITLE
[WIP] Detach diff algorithm from main module [Re-created]

### PR DESCRIPTION
**Pull request by muukii**
## TODO

- [x] Determine module name. => Differentiator
- [x] Determine functions each module.
- [x] Fix Tests.
- [x] Remove WIP commits when merge into master.
- [x] Tidy up Directories
- [ ] Fix podspec
  - [x] Update RxDataSources.podspec
  - [x] Add Differentiator.podspec
  - [ ] `pod trunk push Differentiator.podspec` ( @kzaher please push )

- [x] Bump up version 2.0.0 (Because, this PR has big changes)
- [x] Fix scripts/common.sh
- [ ] Update CHANGELOG.md
- [ ] Update README.md
- [ ] Test `carhtage build --no-skip-current`
- [ ] Test `pod lib lint`
- [x] Fix Travis

## Breaking changes

- Add `Differentiator` Module
- `DifferentiatorError` -> `Diff.Error`
- Remove Example by CocoaPods
- Remove `Differentiator-Test`, Using `Test`
- Remove xcworkspace, Pods/

#149 

## Memo

screen shot 2017-08-05 at 19 10 21

Old #153

### CODE REVIEW

File: ExampleUITests/ExampleUITests.swift [hunk 1]
```diff
@@ -5,29 +5,28 @@
...
-    func testExample() {
-        XCUIApplication().tables.element(boundBy: 0).cells.staticTexts["Randomize Example"].tap()
-
-        let time: TimeInterval = 120.0
-
-        RunLoop.current.run(until: Date().addingTimeInterval(time))
-    }
-    
+
+  override func setUp() {
```
Comment (kzaher) [inline ExampleUITests/ExampleUITests.swift]: I think you have set up tab=2 spaces vs 4 spaces we use. That is causing a lot of noise that aren't really changes.

File: Podfile.lock [hunk 2]
```diff
@@ -1,32 +0,0 @@
-PODS:
```
Comment (kzaher) [inline Podfile.lock]: Why delete this file?

File: Sources/RxDataSources/DataSources/DataSources.swift [hunk 3]
```diff
@@ -8,28 +8,29 @@
 
 import Foundation
 
+@_exported import Differentiator
```
Comment (kzaher) [inline Sources/RxDataSources/DataSources/DataSources.swift]: Is there some reason why we are using `@_exported`?

File: ExampleUITests/ExampleUITests.swift (hunk 1 above)
Comment (muukii) [inline ExampleUITests/ExampleUITests.swift]: Sorry!
I'll fix it!

File: Podfile.lock (hunk 2 above)
Comment (muukii) [inline Podfile.lock]: I've changed this project to Carthage based project.
Therefore, I've removed CocoaPods files.

File: Sources/RxDataSources/DataSources/DataSources.swift (hunk 3 above)
Comment (muukii) [inline Sources/RxDataSources/DataSources/DataSources.swift]: It is indifferent.
But, it helps that keeping only `import RxDataSources` in consumer's codes.

**Comment by muukii** [comment]
@kzaher 
Why did you add RxSwift as git-submodule? (in master-branch)
Vendoring also can by Carthage.
Therefore, Scripts will be getting simpler.

**Comment by kzaher** [comment]
Hi @muukii ,

I think we would like to support CocoaPods, Carthage and git submodules everywhere. SPM is nice to have but I think it still doesn't work for iOS development.

If somebody just wants to hit run in Xcode and check how something works we can just pull RxSwift using submodules and run it. It takes 4 lines of code do do that.

Otherwise somebody maybe doesn't use Cocoapods or Carthage and we'll probably have issues raised that somebody can't run the project, or we won't have issues raised and people will just close it and assume it doesn't work.

We want to have as few assumptions as possible for somebody to be able to successfully run the project once he clones it and once somebody has Xcode, he will have git.

File: Sources/RxDataSources/DataSources/DataSources.swift (hunk 3 above)
Comment (kzaher) [inline Sources/RxDataSources/DataSources/DataSources.swift]: I think that this is a cool trick, but lets just use normal import for now. Differentiator is just an implementation detail.

Comment (kzaher) [inline Sources/RxDataSources/DataSources/DataSources.swift]: I think that this is a cool trick, but lets just use normal import for now. Differentiator is just an implementation detail.

**Pull request merged**